PWGHF/HFC/Tasks/taskCorrelationDsHadrons.cxx

```diff
@@ -523,15 +522,126 @@ struct HfTaskCorrelationDsHadrons {
   }
   PROCESS_SWITCH(HfTaskCorrelationDsHadrons, processMcRecME, "Process MC Reco ME", false);
 
+  SliceCache cache;
+  Preslice<CandDsMcReco> perCollisionCand = o2::aod::hf_cand::collisionId;
+  Preslice<CandDsMcGen> perCollisionCandMc = o2::aod::mcparticle::mcCollisionId;
+
   /// Ds-Hadron correlation - for calculating candidate reconstruction efficiency using MC reco-level analysis
-  void processMcCandEfficiency(soa::Join<aod::Collisions, aod::FT0Mults, aod::EvSels> const&,
+  void processMcCandEfficiency(soa::Join<aod::Collisions, aod::FT0Mults, aod::EvSels, aod::McCollisionLabels> const& collisions,
                                soa::Join<aod::McCollisions, aod::MultsExtraMC> const&,
                                CandDsMcGen const& mcParticles,
                                CandDsMcReco const& candidates,
                                aod::TracksWMc const&)
   {
     auto hCandidates = registry.get<StepTHn>(HIST("hCandidates"));
 
+    /// loop over reconstructed collisions
+    for (const auto& collision : collisions) {
+      // reco collision selection
+      if (useSel8ForEff && !collision.sel8()) {
+        continue;
+      }
+      if (std::abs(collision.posZ()) > cutCollPosZMc) {
+        continue;
+      }
+      if (selNoSameBunchPileUpColl && !(collision.selection_bit(o2::aod::evsel::kNoSameBunchPileup))) {
+        continue;
+      }
+      if (!collision.has_mcCollision()) {
+        registry.fill(HIST("hFakeCollision"), 0.);
+        continue;
+      }
+
+      auto mcCollision = collision.template mcCollision_as<soa::Join<aod::McCollisions, aod::MultsExtraMC>>();
+
+      // TODO: add correctly multiplicity and posZ dependence
+      float multiplicityReco = collision.multFT0M();
+      float multiplicityGen = mcCollision.multMCFT0A() + mcCollision.multMCFT0C(); // multFT0M = multFt0A + multFT0C
+
+      const auto groupedMcParticles = mcParticles.sliceBy(perCollisionCandMc, mcCollision.globalIndex());
+      const auto groupedCandidates = candidates.sliceBy(perCollisionCand, collision.globalIndex());
+
+      // generated candidate loop
+      for (const auto& mcParticle : groupedMcParticles) {
+        if (std::abs(mcParticle.pdgCode()) == Pdg::kDS) {
```

Comment:
Again this is probably not necessary, since it is already included in `(std::abs(mcParticle.flagMcMatchGen()) == 1 << aod::hf_cand_3prong::DecayType::DsToKKPi)`